File: Algorithms/1302_Deepest_Leaves_Sum/Python/Deepest_Leaves_Sum_Solution_1.py

# Space: O(n)
# Time: O(n)

# BFS approach

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def deepestLeavesSum(self, root):
        if root is None: return 0
        if root.left is None and root.right is None: return root.val
        queue = [[root, 0]]
        res, deepest_level = 0, 0
        while queue:
            cur_root, cur_level = queue.pop(0)
            if cur_root.left is None and cur_root.right is None:
                if cur_level == deepest_level:
                    res += cur_root.val
                elif cur_level> deepest_level:
                    res = cur_root.val
                    deepest_level = cur_level

            if cur_root.left: queue.append([cur_root.left,cur_level+1])
            if cur_root.right: queue.append([cur_root.right,cur_level+1])

        return res

File: Algorithms/1302_Deepest_Leaves_Sum/Java/DeepestLeavesSumSolution1.java

// Space: O(n)
// Time: O(n)

// BFS approach

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int deepestLeavesSum(TreeNode root) {
        if (root.left == null && root.right == null) {
            return root.val;
        }
        int res = 0, deepestLevel = 0;
        Queue<MyData> queue = new ArrayDeque<>();
        queue.add(new MyData(root, deepestLevel));

        while (queue.size() > 0) {
            MyData cur = queue.remove();
            if (cur.node.left == null && cur.node.right == null) {
                if (cur.level > deepestLevel) {
                    res = cur.node.val;
                    deepestLevel = cur.level;
                } else if (cur.level == deepestLevel) {
                    res += cur.node.val;
                }
            }
            if (cur.node.left != null) queue.add(new MyData(cur.node.left, cur.level + 1));
            if (cur.node.right != null) queue.add(new MyData(cur.node.right, cur.level + 1));
        }
        return res;

    }

    class MyData {
        public TreeNode node;
        public int level;

        public MyData(TreeNode node, int level) {
            this.node = node;
            this.level = level;
        }
    }
}




